chain/network/src/routing/routing_table_actor.rs

```diff
@@ -385,6 +385,10 @@ impl RoutingTableActor {
 
         res
     }
+
+    pub fn get_all_edges(&self) -> Vec<Edge> {
+        self.edges_info.iter().map(|x| x.1.clone()).collect()
```

Comment:
if @matklad  agrees - I'd go ahead and change edges_info to have this Arc within this PR

Reply:
I think it's fine! Perhaps a bit more rustic design would be to represent both peers and edges as just integer indices into a table which contains the actual data, but managing such a table between two actors would be a pain. 

Given that size_of edge is a whopping `344` bytes, I wonder if we should even move the Arc inside the edge itself? 

```rust
pub struct Edge {
    repr: Arc<EdgeRepr>
}

struct EdgeRepr {
    peer0: PeerId,
    peer1: PeerId,
    ...
}
```

Keeping such large objects on the stack seems inefficient. 

Anyway, just sticking an Arc there by itself is :+1:

Reply:
@matklad I was able to further optimize the code changing `edge_info` to be `HashMap<Arc<PeerId, PeerId>, Arc<Edge>>`
and `Edge` to be
```rust

pub struct EdgeNew {.
    pub key: Arc<(PeerId, PeerId)>,
    ...
}
```

There are a few reasons to use this structure:
- performance of cloning -This helped reduce the time to clone Edges from `1.25s to 0.95s`. It is really inefficient to have large structures as keys / values in hashmap.

- memory usage per item - `(PeerId, PeerId)` is around 132 bytes, and edge is 344 bytes. That given total of around 500 bytes. Given that in hashmap your occupancy has to  be between 25% to 50%, in the worst case you will have to use 4x as much memory. So you end up using around 2kb per Edge.

Assuming we have `n` edges.
Each element of hashmap would be 16 bytes.
With the new approach: in the worst case we will be using `16*4*n +  (344 + 8) * n =  (64 + 344 + 8) * n = (416) * n`. That is 416 bytes per item.
With old one it would be: `(132+344)*4*n = 1904 * n`. So `1904` bytes per each item.

- In addition we are using `(PeerId, PeerId)` copy twice for each edge  in `SlotMap` inside `ibf_peer_set.rs`. So we can reduce the number of allocations needed.

- We can go even further and reduce duplicates of `PeerId`. We can achieve this by creating `HashSet<Arc<PeerId>>`. And then replacing `PeerId` with a global copy of `Arc<PeerId>`,

Anyway, to be careful, we could make 3 separate changes:
- replace Edge with Arc<Edge>
- replace `key: <(PeerId, PeerId)>,` with `key: Arc<(PeerId, PeerId)>,`
- replace `PeerId` with `Arc<PeerId>`

Reply:
> Given that size_of edge is a whopping `344` bytes, I wonder if we should even move the Arc inside the edge itself?

Let's do changes step by step. That will allow us to measure benefits vs having to add extra `Arc<...>`.
I think it makes sense to use `Arc<...>` on all structures that are at least 64, maybe even 32 bytes, that we use as values or keys inside `HashMap` / `HashSet`.

Reply:
>Given that in hashmap your occupancy has to be between 25% to 50%,

I think Rust's hashmap is build for much higher occupancy: 

https://github.com/rust-lang/hashbrown/blob/462666a0bc6738793912c133878e533dbf478b0d/src/raw/mod.rs#L203

>`HashMap<Arc<PeerId, PeerId>, Arc<Edge>>`

I still don't like that we duplicate the key twice here (as peer pair is already stored as part of an Edge), Id try to refactor that to be represented as a has set internally. THough just arcing the key is ok as well probably.

Reply:
> > Given that in hashmap your occupancy has to be between 25% to 50%,
> 
> I think Rust's hashmap is build for much higher occupancy:
> 
> https://github.com/rust-lang/hashbrown/blob/462666a0bc6738793912c133878e533dbf478b0d/src/raw/mod.rs#L203
> 
> > `HashMap<Arc<PeerId, PeerId>, Arc<Edge>>`
> 
> I still don't like that we duplicate the key twice here (as peer pair is already stored as part of an Edge), Id try to refactor that to be represented as a has set internally. THough just arcing the key is ok as well probably.

We can make a separate PR for replacing `edges_info` with `HashMap<Arc<PeerId>, Arc<PeerId>, Arc<Edge>>`.
However, first we should do other refactoring. I checked that there wasn't much performance difference between those two.

Btw, `Arc` is adding extra `16` bytes to each allocation. Maybe, we should work `Arc` and only keep `Strong` as `AtomicUsize32`. That would reduce overhead to only 4 bytes.